File: truevfs-kernel/src/main/java/de/schlichtherle/truevfs/kernel/se/ClutchInputSocket.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se;

import java.io.IOException;
import java.io.InputStream;
import java.nio.channels.SeekableByteChannel;
import javax.annotation.concurrent.NotThreadSafe;
import net.truevfs.kernel.cio.DecoratingInputSocket;
import net.truevfs.kernel.cio.Entry;
import net.truevfs.kernel.cio.InputSocket;

/**
 * An input lazySocket which obtains its delegate lazySocket lazily and
 * {@link #reset()}s it upon any {@link Throwable}.
 *
 * @see    ClutchOutputSocket
 * @param  <E> the type of the {@link #localTarget() local target}.
 * @author Christian Schlichtherle
 */
@NotThreadSafe
abstract class ClutchInputSocket<E extends Entry>
extends DecoratingInputSocket<E> {

    @Override
    protected final InputSocket<? extends E> socket() throws IOException {
        final InputSocket<? extends E> socket = this.socket;
        return null != socket ? socket : (this.socket = lazySocket());
    };

    /**
     * Returns the input lazySocket for lazy initialization.
     * 
     * @return the input lazySocket for lazy initialization.
     * @throws IOException on any I/O error. 
     */
    protected abstract InputSocket<? extends E> lazySocket()
    throws IOException;

    @Override
    public E localTarget() throws IOException {
        try {
            return boundSocket().localTarget();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    @Override
    public InputStream stream() throws IOException {
        try {
            return boundSocket().stream();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    @Override
    public SeekableByteChannel channel()
    throws IOException {
        try {
            return boundSocket().channel();
        } catch (final Throwable ex) {
            reset();
            throw ex;
        }
    }

    protected final void reset() {
        this.socket = null;
    }
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/ClutchInputSocket.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel

import javax.annotation.concurrent._
import net.truevfs.kernel.cio._
import scala.annotation.tailrec

/** An input socket which obtains its delegate socket lazily and `reset`s
  * it upon any [[java.lang.Throwable]].
  *
  * @see    ClutchOutputSocket
  * @tparam E the type of the `localTarget`.
  * @author Christian Schlichtherle
  */
@NotThreadSafe
private abstract class ClutchInputSocket[E <: Entry]
extends DelegatingInputSocket[E] {

  private[this] var _socket: Option[InputSocket[_ <: E]] = None

  override def localTarget() = apply(_.localTarget())

  override def stream() = apply(_.stream())

  override def channel() = apply(_.channel())

  private def apply[A](f: InputSocket[_ <: E] => A) = {
    try { f(boundSocket) }
    catch { case ex => reset(); throw ex }
  }

  protected final def reset() { _socket = None }

  @tailrec
  protected final override def socket: InputSocket[_ <: E] = {
    _socket match {
      case Some(socket) => socket
      // In case lazySocket returns null, this will produce an NPE, which is
      // better than a retry which may probably only result in an endless loop.
      case None => _socket = Some(lazySocket); socket
    }
  }

  /** Returns the input socket for lazy initialization.
    * 
    * @return the input socket for lazy initialization.
    * @throws IOException on any I/O error. 
    */
  protected def lazySocket: InputSocket[_ <: E]

  /** Returns a string representation of this object for debugging and logging
    * purposes.
    */
  override def toString = "%s[socket=%s]" format (getClass.getName, socket)
}
